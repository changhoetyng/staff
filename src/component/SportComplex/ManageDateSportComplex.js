import React, { Component } from "react";
import Header from "../../sharedComponent/Header";
import Table from "react-bootstrap/Table";
import DropdownButton from "react-bootstrap/DropdownButton";
import Dropdown from "react-bootstrap/Dropdown";
import DatePicker from "react-datepicker";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import "react-datepicker/dist/react-datepicker.css";
import moment from "moment";
import Button from "react-bootstrap/Button";
import Modal from "react-bootstrap/Modal";
import { api } from "../../api/api";

class SportComplex extends Component {
  constructor(props) {
    super(props);
    this.state = {
      history: this.props.history,
      currentSelection: null,
      currentFacilityId: null,
      selectedSub: null,
      selectedSubId: null,
      date: null,
      checkedItems: new Map(),
      activeData: { time: "", timeStatus: { status: "", studentId: "" } },
      modalOn: false,
      data: [],
    };
  }

  async componentDidMount() {
    this.setState({ loading: true });
    await api
      .get("/sportComplex/getFacility")
      .then((res) => {
        console.log(res.data.data);
        this.setState({ data: res.data.data });
      })
      .catch((err) => {
        console.log(err);
      });
    this.setState({ loading: false });
  }

  dropdownSelection(selected, id) {
    this.setState({ currentSelection: selected, currentFacilityId: id });
  }

  dropdownSub(selected, id) {
    this.setState({ selectedSub: selected, selectedSubId: id });
  }

  renderFacilityButton() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          this.state.currentSelection
            ? this.state.currentSelection
            : "Select a facility"
        }
      >
        {this.state.data.map((v, i) => {
          return (
            <Dropdown.Item
              key={i}
              as="button"
              onSelect={() => this.dropdownSelection(v.facility, v.facilityId)}
            >
              {v.facility}
            </Dropdown.Item>
          );
        })}
      </DropdownButton>
    );
  }

  renderCategoryButton() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          !this.state.currentSelection
            ? "Please Select a Facility"
            : this.state.selectedSub
            ? this.state.selectedSub
            : "Select a subcategory"
        }
      >
        {this.state.data.map((v, i) => {
          return (
            v.facilityId === this.state.currentFacilityId && v.subCategory &&
                v.subCategory.map((sub,j) => {
                  return (
                    <Dropdown.Item
                      key={j}
                      as="button"
                      onSelect={() =>
                        this.dropdownSub(sub.subName, sub._id)
                      }
                    >
                      {sub.subName}
                    </Dropdown.Item>
                  );
                })
            
          );
        })}
      </DropdownButton>
    );
  }

  handleChange = (e) => {
    const item = e.target.name;
    const isChecked = e.target.checked;
    this.setState((prevState) => ({
      checkedItems: prevState.checkedItems.set(item, isChecked),
    }));
  };

  renderTableContent() {
    let selectedFacility = this.state.data.find(
      (v) => v.facilityId === this.state.currentFacilityId 
    );

    let selectedDate = selectedFacility
      ? selectedFacility.data.find(
          (v) => v.date === moment(this.state.date).format("DD/MM/YYYY") && v.subCategoryId === this.state.selectedSubId
        )
      : null;
    let timeData = selectedDate ? selectedDate.timeListing : [];
    return (
      <tbody>
        {timeData.map((v, i) => {
          return (
            <tr key={i}>
              <td>{v.time}</td>
              <td>{v.timeStatus.status}</td>
              <td>
                <Button
                  variant="link"
                  style={{ marginLeft: 10 }}
                  onClick={() =>
                    this.setState({ activeData: v, modalOn: true })
                  }
                >
                  More Details
                </Button>
              </td>
            </tr>
          );
        })}
      </tbody>
    );
  }

  hideModal() {
    this.setState({ modalOn: false });
  }

  async openBooking() {
    this.setState({ loading: true });
    await api
      .patch("/sportComplex/openTime", {
        facilityId: this.state.currentFacilityId,
        date: moment(this.state.date).format("DD/MM/YYYY"),
        time: this.state.activeData.time,
        subCategoryId: this.state.selectedSubId
      })
      .catch((err) => {
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, modalOn: false });
  }

  async closeBooking() {
    this.setState({ loading: true });
    await api
      .patch("/sportComplex/closeTime", {
        facilityId: this.state.currentFacilityId,
        date: moment(this.state.date).format("DD/MM/YYYY"),
        time: this.state.activeData.time,
        subCategoryId: this.state.selectedSubId,
      })
      .catch((err) => {
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, modalOn: false });
  }

  async cancelBooking() {
    this.setState({ loading: true });
    await api
      .patch("/sportComplex/cancelBooking", {
        facilityId: this.state.currentFacilityId,
        date: moment(this.state.date).format("DD/MM/YYYY"),
        time: this.state.activeData.time,
        subCategoryId: this.state.selectedSubId,
        bookingId: this.state.activeData.timeStatus.bookingId,
        studentId: this.state.activeData.timeStatus.studentId
      })
      .catch((err) => {
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, modalOn: false });
  }

  render() {
    return (
      <div>
        <Header history={this.state.history} />
        <Modal
          show={this.state.modalOn}
          onHide={() => this.hideModal()}
          centered
        >
          <Modal.Header>
            <Modal.Title>Details</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            Date: {moment(this.state.date).format("DD/MM/YYYY")}
            <br />
            Time: {this.state.activeData.time}
            <br />
            Status: {this.state.activeData.timeStatus.status}
            <br />
            {this.state.activeData.timeStatus.status === "booked" && (
              <div>
                Student ID: {this.state.activeData.timeStatus.studentId}
              </div>
            )}
          </Modal.Body>
          <Modal.Footer>
            {this.state.activeData.timeStatus.status === "booked" && (
              <Button variant="danger" onClick={() => this.cancelBooking()}>
                Cancel booking
              </Button>
            )}
            {this.state.activeData.timeStatus.status === "open" && (
              <Button id="closeBooking" onClick={() => this.closeBooking()}>Close booking</Button>
            )}
            {this.state.activeData.timeStatus.status === "close" && (
              <Button id="openBooking" onClick={() => this.openBooking()}>Open booking</Button>
            )}
          </Modal.Footer>
        </Modal>
        <div className="container float-left">
          <div className="container-fluid" style={{ marginTop: 20 }}>
            <h4 id="title">Sport Complex</h4>
          </div>
          <div
            className="container-fluid"
            style={{ marginTop: 20, flexDirection: "row" }}
          >
            <Row>
              <Col>{this.renderFacilityButton()}</Col>
              <Col>{this.renderCategoryButton()}</Col>
              {
                <Col>
                  <DatePicker
                    selected={this.state.date}
                    isClearable
                    onChange={(date) => this.setState({ date })}
                  />
                </Col>
              }
            </Row>
          </div>
          <div className="container-fluid" style={{ marginTop: 20 }}>
            <Table striped bordered hover>
              <thead>
                <tr>
                  <th>Time</th>
                  <th>Status</th>
                  <th>Details</th>
                </tr>
              </thead>
              {this.renderTableContent()}
            </Table>
          </div>
        </div>
      </div>
    );
  }
}

export default SportComplex;
