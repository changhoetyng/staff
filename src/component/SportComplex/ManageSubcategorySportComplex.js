import React, { Component } from "react";
import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import Dropdown from "react-bootstrap/Dropdown";
import DropdownButton from "react-bootstrap/DropdownButton";
import FullPageLoader from "../../hooks/FullPageLoader";
import Header from "../../sharedComponent/Header";
import "../../css/Setting.css";
import { api } from "../../api/api";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Modal from "react-bootstrap/Modal";
import { QRCode } from "react-qr-svg";

class ManageSubcategorySportComplex extends Component {
  constructor(props) {
    super(props);
    this.state = {
      history: this.props.history,
      currentSelection: null,
      currentFacilityId: null,
      subCategory: "",
      data: [],
      loading: false,
      qrCodeModal: false,
      selectedQR: "",
      isAdmin: false,
    };
  }

  async componentDidMount() {
    await api
      .get("/user")
      .then((res) => {
        const role = res.data.user.role;
        if (role === "admin") {
          this.setState({ isAdmin: true });
        }
      })
      .catch((err) => console.log(err));

    this.setState({ loading: true });
    await api
      .get("/sportComplex/getFacility")
      .then((res) => {
        this.setState({ data: res.data.data });
      })
      .catch((err) => {
        console.log(err);
      });
    this.setState({ loading: false });
  }

  dropdownSelection(selected, id) {
    this.setState({ currentSelection: selected, currentFacilityId: id });
  }

  openQrCode(sub) {
    console.log(sub);
    this.setState({ selectedQR: sub, qrCodeModal: true });
  }

  renderFacilityButton() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          this.state.currentSelection
            ? this.state.currentSelection
            : "Select a facility"
        }
      >
        {this.state.data.map((v, i) => {
          return (
            <Dropdown.Item
              key={i}
              as="button"
              onSelect={() => this.dropdownSelection(v.facility, v.facilityId)}
            >
              {v.facility}
            </Dropdown.Item>
          );
        })}
      </DropdownButton>
    );
  }

  renderTableContent() {
    let selectedFacility = this.state.data.find(
      (v) => v.facilityId === this.state.currentFacilityId
    );
    let subCategory = selectedFacility ? selectedFacility.subCategory : [];
    return (
      <div className="row">
        {subCategory &&
          subCategory.map((sub, i) => {
            return (
              <div className="col" key={i}>
                <Card
                  style={{
                    width: "19rem",
                    whiteSpace: "pre-wrap",
                    paddingRight: 0,
                  }}
                >
                  <Card.Body>
                    <Card.Title>{sub.subName}</Card.Title>
                    <Card.Text>
                      {sub.currentUser ? (
                        <p>Current User: {sub.currentUser}</p>
                      ) : (
                        <p>Current User: Not occupied</p>
                      )}
                    </Card.Text>
                    <Card.Link
                      style={{ cursor: "pointer" }}
                      onClick={() => {
                        this.openQrCode(sub);
                      }}
                    >
                      QR Code
                    </Card.Link>
                    {this.state.isAdmin && (
                      <Card.Link
                        style={{ cursor: "pointer" }}
                        onClick={() => {
                          this.removesubcategory(sub._id);
                        }}
                      >
                        Remove
                      </Card.Link>
                    )}
                  </Card.Body>
                </Card>
              </div>
            );
          })}
      </div>
    );
  }

  async removesubcategory(subId) {
    this.setState({ loading: true });

    await api
      .patch("/sportComplex/deletesub", {
        facilityId: this.state.currentFacilityId,
        subCategoryId: subId,
      })
      .catch((err) => {
        alert(err);
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false });
  }

  async addsubcategory() {
    this.setState({ loading: true });

    await api
      .patch("/sportComplex/addSub", {
        facilityId: this.state.currentFacilityId,
        subCategory: this.state.subCategory,
      })
      .catch((err) => {
        alert(err);
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, subCategory: "" });
  }

  handleClose() {
    this.setState({ qrCodeModal: false });
  }

  render() {
    return (
      <div>
        {this.state.loading && <FullPageLoader />}
        <Header history={this.state.history} />
        <div className="container float-left non-printable">
          <h2>Manage Subcategory</h2>

          <div style={{ marginTop: 20 }}>
            <h4>Sport Complex</h4>
            <div className="container float-left">
              <div className="row" style={{ marginTop: 20 }}>
                {this.renderFacilityButton()}
              </div>
              <div className="row" style={{ marginBottom: 20, marginTop: 20 }}>
                {this.renderTableContent()}
              </div>
              {this.state.isAdmin && (
                <div className="row,container float-left">
                  <div className="row">
                    <p className="word">Manage Subcategory:</p>
                    <div className="col">
                      <input
                        type="text"
                        value={this.state.subCategory}
                        onChange={(e) =>
                          this.setState({ subCategory: e.target.value })
                        }
                      />
                    </div>
                    <div className="col">
                      <Button onClick={() => this.addsubcategory()}>Add</Button>
                    </div>
                  </div>
                </div>
              )}
            </div>
          </div>
        </div>
        <div class="non-printable">
          <Modal
            show={this.state.qrCodeModal}
            aria-labelledby="contained-modal-title-vcenter"
            onHide={() => this.handleClose()}
            centered
          >
            <Modal.Header>
              <Modal.Title>QR Code</Modal.Title>
            </Modal.Header>
            <Modal.Body>
              <div class="printable">
                <h2 style={{ textAlign: "center" }}>
                  {this.state.currentSelection}
                </h2>
                <h4 style={{ textAlign: "center" }}>
                  {this.state.selectedQR.subName}
                </h4>
                <div style={{ textAlign: "center" }}>
                  <QRCode
                    level="Q"
                    style={{ width: 256 }}
                    value={JSON.stringify({
                      venueId: this.state.currentFacilityId,
                      subCategoryId: this.state.selectedQR._id,
                    })}
                  />
                </div>
              </div>
            </Modal.Body>
            <Modal.Footer>
              <Button onClick={() => window.print()}>Print</Button>
            </Modal.Footer>
          </Modal>
        </div>
      </div>
    );
  }
}

export default ManageSubcategorySportComplex;
