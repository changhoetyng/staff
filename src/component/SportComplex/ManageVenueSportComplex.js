import React, { Component } from "react";
import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import FullPageLoader from "../../hooks/FullPageLoader";
import DatePicker from "react-datepicker";
import Header from "../../sharedComponent/Header";
import "../../css/Setting.css";
import { api } from "../../api/api";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Dropdown from "react-bootstrap/Dropdown";
import Modal from "react-bootstrap/Modal";
import DropdownButton from "react-bootstrap/DropdownButton";
import moment from "moment";

class ManageVenue extends Component {
  constructor(props) {
    super(props);
    this.state = {
      history: this.props.history,
      currentSelection: null,
      selectedSubCategory: null,
      data: [],
      loading: false,
      fields: {},
      isModal: false,
      selectedSubCategoryId: null,
      error: "",
      facility: "",
      isAdmin: false
    };
  }

  async componentDidMount() {
    this.setState({ loading: true });
    await api
      .get("/user")
      .then((res) => {
        const role = res.data.user.role
        if(role === "admin"){
          this.setState({isAdmin: true})
        }
      }).catch((err) => console.log(err))
    
    await api
      .get("/sportComplex/getFacility")
      .then((res) => {
        this.setState({ data: res.data.data });
      })
      .catch((err) => {
        console.log(err);
      });
    this.setState({ loading: false });
  }

  handleDateChange = (dateName, dateValue) => {
    this.setState({
      fields: {
        ...this.state.fields,
        [dateName]: dateValue,
      },
    });
  };

  renderDropdown() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          this.state.selectedSubCategory
            ? this.state.selectedSubCategory
            : "Select"
        }
      >
        {this.state.currentSelection &&
          this.state.currentSelection.subCategory &&
          this.state.currentSelection.subCategory.map((v, i) => {
            return (
              <Dropdown.Item
                key={i}
                as="button"
                onSelect={() =>
                  this.setState({
                    selectedSubCategoryId: v._id,
                    selectedSubCategory: v.subName,
                  })
                }
              >
                {v.subName}
              </Dropdown.Item>
            );
          })}
      </DropdownButton>
    );
  }

  renderSportComplexCard() {
    return this.state.data.map((data, index) => {
      return (
        <Col key={index}>
          <Card className="text-center" style={{ width: 300 }}>
            <Card.Body>
              <Card.Title>{data.facility}</Card.Title>
              <DatePicker
                selected={this.state.fields[data.facilityId]}
                minDate={new Date()}
                isClearable
                onChange={(date) =>
                  this.handleDateChange(data.facilityId, date)
                }
              />
            </Card.Body>
            {this.state.fields[data.facilityId] && (
              <Card.Body>
                <Card.Link
                  style={{ cursor: "pointer" }}
                  onClick={() =>
                    this.setState({ currentSelection: data, isModal: true })
                  }
                >
                  Open Date
                </Card.Link>
              </Card.Body>
            )}
            { this.state.isAdmin &&
            <Card.Body>
              <Card.Link
                style={{ cursor: "pointer" }}
                onClick={() =>
                  this.removeFacility(data.facilityId)
                }
              >
                Remove
              </Card.Link>
            </Card.Body>
            }
          </Card>
        </Col>
      );
    });
  }

  async removeFacility(facilityId){
    this.setState({ loading: true });
    await api
      .delete("/sportComplex/deleteFacility", {
        data: {facilityId}
      })
      .then(() => {
        this.componentDidMount()
      })
      .catch((err) => {
       console.log(err.response)
      });
    this.setState({ loading: false });
  }

  async openDate() {
    const facilityId = this.state.currentSelection.facilityId;
    const subCategoryId = this.state.selectedSubCategoryId;
    const date = moment(this.state.fields[facilityId]).format("DD/MM/YYYY");
    this.setState({ loading: true });
    await api
      .post("/sportComplex/openDate", {
        facilityId,
        date,
        subCategoryId,
      })
      .then(() => {
        this.setState({ isModal: false });
      })
      .catch((err) => {
        if (err.response.data.message === "time already exists") {
          this.setState({ error: "Time already existed" });
        }
      });
    this.setState({ loading: false });
  }

  handleClose() {
    this.setState({
      isModal: false,
      selectedSubCategory: null,
      selectedSubCategoryId: null,
      currentSelection: null,
      error: "",
    });
  }

  async addFacility() {
    this.setState({ loading: true });
    await api
      .post("/sportComplex/addFacility", {
        name: this.state.facility,
      })
      .catch((err) => {
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false });
  }

  render() {
    return (
      <div>
        <Modal
          show={this.state.isModal}
          aria-labelledby="contained-modal-title-vcenter"
          onHide={() => this.handleClose()}
          centered
        >
          <Modal.Body className="text-center">
            <h4>Select subcategory</h4>
            <div>{this.renderDropdown()}</div>
            {this.state.selectedSubCategoryId && (
              <Button onClick={() => this.openDate()}>Open</Button>
            )}
            <p style={{ color: "red", alignSelf: "left" }}>
              {this.state.error}
            </p>
          </Modal.Body>
        </Modal>
        {this.state.loading && <FullPageLoader />}
        <Header history={this.state.history} />
        <div className="container float-left">
          <h2 id="manageVenueTitle">Manage Venue</h2>
          <div style={{ marginTop: 20 }}>
            <h4 id="title">Sport Complex</h4>
          </div>
          <div style={{ marginTop: 20 }}>
            <Row>{this.renderSportComplexCard()}</Row>
          </div>
          {
            this.state.isAdmin && 
            <div className="container float-left mt-4">
            <div className="row">
              <p className="word">Add Facility:</p>
              <div className="col">
                <input
                  type="text"
                  value={this.state.subCategory}
                  onChange={(e) => this.setState({ facility: e.target.value })}
                />
                <Button
                  style={{ marginLeft: 20 }}
                  onClick={() => this.addFacility()}
                >
                  Add
                </Button>
              </div>
            </div>
          </div>
          }
          
        </div>
      </div>
    );
  }
}

export default ManageVenue;
