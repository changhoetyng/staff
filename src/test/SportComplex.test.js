import React from "react";
import { configure, shallow } from "enzyme";

// Import Page
import ManageDateSportComplex from "../component/SportComplex/ManageDateSportComplex";
import ManageSubcategorySportComplex from "../component/SportComplex/ManageSubcategorySportComplex";
import ManageVenueSportComplex from "../component/SportComplex/ManageVenueSportComplex";

describe("ManageDateSportComplex", () => {
  const manageDateSportComplex = shallow(<ManageDateSportComplex />);

  test("Title should contain word \"Sport Complex\"", () => {
    const title = manageDateSportComplex.find("#title");
    expect(title.text()).toBe("Sport Complex");
  });

  test("Snapshot ManageDateSportComplex", () => {
    expect(manageDateSportComplex).toMatchSnapshot();
  });
});

describe("ManageSubcategorySportComplex", () => {
  const manageSubcategorySportComplex = shallow(
    <ManageSubcategorySportComplex />
  );

  test("Title should contain word \"Sport Complex\"", () => {
    const title = manageSubcategorySportComplex.find("#title");
    expect(title.text()).toBe("Sport Complex");
  });

  test("manageSubcategoryTitle should contain word \"Manage Subcategory\"", () => {
    const title = manageSubcategorySportComplex.find("#manageSubcategoryTitle");
    expect(title.text()).toBe("Manage Subcategory");
  });

  test("Snapshot ManageSubcategorySportComplex", () => {
    expect(manageSubcategorySportComplex).toMatchSnapshot();
  });
});

describe("ManageVenueSportComplex", () => {
  const manageVenueSportComplex = shallow(<ManageVenueSportComplex />);

  test("change password button should activate modal", () => {
    const title = manageVenueSportComplex.find("#title");
    expect(title.text()).toBe("Sport Complex");
  });

  test("Title should contain word \"Manage Subcategory\"", () => {
    const title = manageVenueSportComplex.find("#manageVenueTitle");
    expect(title.text()).toBe("Manage Venue");
  });

  test("Snapshot ManageVenueSportComplex", () => {
    expect(manageVenueSportComplex).toMatchSnapshot();
  });
});
