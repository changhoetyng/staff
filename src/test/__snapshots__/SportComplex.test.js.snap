// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageDateSportComplex Snapshot ManageDateSportComplex 1`] = `ShallowWrapper {}`;

exports[`ManageSubcategorySportComplex Snapshot ManageSubcategorySportComplex 1`] = `ShallowWrapper {}`;

exports[`ManageVenueSportComplex Snapshot ManageVenueSportComplex 1`] = `ShallowWrapper {}`;
